### PR TITLE
feat: support function context prompt

### PR DESCRIPTION
**Pull request by bitaoz**
**Summary**
Add support for function context prompt after entering a function name. See https://github.com/sassoftware/vscode-sas-extension/issues/66 for more details.

**Testing**
Tested entering a normal sas function.
Tested entering a nested sas function within another sas function.
Tested entering a sas macro function.

### CODE REVIEW

File: server/src/sas/CompletionProvider.ts [hunk 1]
```diff
@@ -281,6 +284,381 @@ export class CompletionProvider {
...
+                      }
+                      _arguments = _arguments.map(
+                        (_argument) =>
+                          _argument
+                            ?.replace("(s)", "")
+                            .replace(/['"‘’<>()|.]/g, "")
+                            .replace(/[-_\s]?\d$/, "")
+                            .replace(/[-_\s]n$/, "")
+                            .trim()
+                            .toLowerCase(),
```
Comment (scnwwu) [inline server/src/sas/CompletionProvider.ts:472]: looks like used at several places, consider extract to a function like `cleanupArg`

File: server/src/sas/CompletionProvider.ts [hunk 2]
```diff
@@ -281,6 +284,381 @@ export class CompletionProvider {
...
+                        .filter((item) => item);
+
+                      if (_arguments.length === 1) {
+                        return [
+                          argument.includes("(s)")
+                            ? argument.replace(/[<>]/g, "").trim()
+                            : argument.replace(/[<>()]/g, "").trim(),
+                          descriptions[0] ?? "",
+                          true,
+                        ] as [string, string, boolean];
```
Comment (scnwwu) [inline server/src/sas/CompletionProvider.ts:533]: avoid using `as` as long as possible.
declare the type as `.map<[string, string, boolean]>()`

File: server/src/sas/CompletionProvider.ts [hunk 3]
```diff
@@ -281,6 +284,381 @@ export class CompletionProvider {
...
+                  documentation: data.data,
+                  parameters: argumentsInSyntax?.map((item) => {
+                    if (
+                      item[0].replace(/['"‘’<>()|]/g, "") === "..." ||
+                      item[0].replace(/['"‘’<>()|]/g, "") === "…"
+                    ) {
+                      return {
+                        label: item[0],
+                        documentation: {
+                          kind: MarkupKind.Markdown,
```
Comment (scnwwu) [inline server/src/sas/CompletionProvider.ts:617]: do not duplicate these 3 lines, simply calculate different documentation value

File: server/src/sas/CompletionProvider.ts [hunk 4]
```diff
@@ -281,6 +284,381 @@ export class CompletionProvider {
...
+                    leftArgumentsDocumentation = [
+                      ...new Set(leftArguments),
+                    ].join("\n\n");
+                  }
+                }
+
+                signatures.push({
+                  label: syntax,
+                  documentation: data.data,
+                  parameters: argumentsInSyntax?.map((item) => {
```
Comment (scnwwu) [inline server/src/sas/CompletionProvider.ts:609]: deconstruct the `item` to be self explained, for example `([name, description, isMatch]) => {`

File: server/src/sas/CompletionProvider.ts [hunk 5]
```diff
@@ -281,6 +284,291 @@ export class CompletionProvider {
...
+                arg = arg
+                  .replace("(s)", "")
+                  .replace(/['"‘’<>()|.]/g, "")
+                  .replace(/[-_\s]?\d$/, "")
+                  .replace(/[-_]n$/, "")
+                  .trim()
+                  .toLowerCase();
+                if (arg.match(/\w+\s+n$/)) {
+                  return arg.replace(/\s+n$/, "");
+                }
```
Comment (scnwwu) [inline server/src/sas/CompletionProvider.ts:386]:
```suggestion
                arg.replace(/\b\s+n$/, "");
```

File: server/src/sas/CompletionProvider.ts [hunk 6]
```diff
@@ -281,6 +295,277 @@ export class CompletionProvider {
...
+                        : unmatchedArgsDocumentation
+                        ? outerArgsDocumentation
+                          ? `${unmatchedArgsDocumentation}\n\n${outerArgsDocumentation}`
+                          : unmatchedArgsDocumentation
+                        : outerArgsDocumentation
+                        ? outerArgsDocumentation
+                        : "";
+                      const documentLink = `${getText(
+                        "ce_ac_sas_function_doc_txt",
+                      )}: ${docLink}`;
```
Comment (scnwwu) [inline server/src/sas/CompletionProvider.ts:541]: colon needs to be part of externalized message

**Review by scnwwu — APPROVED** [review]

**Comment by Zhirong2022** [comment]
Function btrim()
'btrim-character' is repeated on function help page.
image

**Comment by Zhirong2022** [comment]
Function %sysfunc()
The first parameter is not highlighted with typing.
image

**Comment by Zhirong2022** [comment]
Function git_status_get()
The help contents for parameter 'attribute' is not shown as expected.
image

**Comment by Zhirong2022** [comment]
Function kpropcase()
Additional symbol star on syntax help.
image

**Comment by Zhirong2022** [comment]
Function tslvl()
The option values are not fully listed on function help page.
image

**Pull request merged**